File: 123-best-time-to-buy-and-sell-stock-iii/best-time-to-buy-and-sell-stock-iii.py
class Solution:
    def maxProfit(self, price: List[int]) -> int:
        n=len(price)
        dp={}
        def rec(i,buy,cnt):
            if i>=n or  cnt>2:
                return 0
            temp=str(buy)+"_"+str(cnt)
            if (i,temp) in dp:
                return dp[(i,temp)]
            if not buy:
                profit=max( -price[i]+rec(i+1,1,cnt+1) , rec(i+1,0,cnt))
            else:
                profit=max(price[i]+rec(i+1,0,cnt), rec(i+1,buy,cnt))
            
            dp[(i,temp)]=profit
            return profit
        
        return rec(0,0,0)
            
        

File: 123-best-time-to-buy-and-sell-stock-iii/best-time-to-buy-and-sell-stock-iii.cpp
int rec(int idx,vector<int>&prices,int prev_buy,vector<vector<vector<int>>>&dp,int trans){


    if(idx==prices.size()||trans==2)return 0;

    if(dp[idx][prev_buy][trans]!=-1)return dp[idx][prev_buy][trans];
    if(prev_buy ){
     
        
        int sell=prices[idx]+rec(idx+1,prices,0,dp,trans+1);
        int rest=rec(idx+1,prices,prev_buy,dp,trans);
        return dp[idx][prev_buy][trans]= max(sell,rest);
        // profit=max({profit,sell,rest});
    }
  
        int buy=-prices[idx]+rec(idx+1,prices,1,dp,trans);
        int rest=rec(idx+1,prices,prev_buy,dp,trans);
        return dp[idx][prev_buy][trans]=max(buy,rest);


    

}



class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int n=prices.size();
        vector<vector<vector<int>>>dp(n,vector<vector<int>>(2,vector<int>(3,-1)));
        return rec(0,prices,0,dp,0);
    }
};